crates/apps_lib/src/client/masp.rs

```diff
@@ -28,137 +30,209 @@ pub async fn syncing<
     sks: &[ExtendedSpendingKey],
     fvks: &[ViewingKey],
 ) -> Result<ShieldedContext<U>, Error> {
-    let shutdown_signal = async {
-        let (tx, rx) = tokio::sync::oneshot::channel();
-        namada_sdk::control_flow::shutdown_send(tx).await;
-        rx.await
-    };
-
     display_line!(io, "{}", "==== Shielded sync started ====".on_white());
     display_line!(io, "\n\n");
-    let logger = CliLogger::new(io);
-    let sync = async move {
-        shielded
-            .fetch(
-                client,
-                &logger,
-                start_query_height,
-                last_query_height,
-                batch_size,
-                sks,
-                fvks,
-            )
-            .await
-            .map(|_| shielded)
-    };
-    tokio::select! {
-        sync = sync => {
-            let shielded = sync?;
-            display!(io, "Syncing finished\n");
-            Ok(shielded)
-        },
-        sig = shutdown_signal => {
-            sig.map_err(|e| Error::Other(e.to_string()))?;
-            display!(io, "\n");
-            Ok(ShieldedContext::default())
-        },
-    }
+    let tracker = CliProgressTracker::new(io);
+
+    let shielded = shielded
+        .fetch(
+            LedgerMaspClient::new(client),
+            &tracker,
+            start_query_height,
+            last_query_height,
+            RetryStrategy::Forever,
+            batch_size,
+            sks,
+            fvks,
+        )
+        .await
+        .map(|_| shielded)?;
+
+    display!(io, "Syncing finished\n");
+    Ok(shielded)
 }
 
-pub struct CliLogging<'io, T, IO: Io> {
-    items: Vec<T>,
+/// The amount of progress a shielded sync sub-process has made
+#[derive(Default, Copy, Clone, Debug)]
+struct IterProgress {
     index: usize,
     length: usize,
+}
+
+pub struct LoggingIterator<'io, T, I, IO>
+where
+    T: Debug,
+    I: Iterator<Item = T>,
+    IO: Io,
+{
+    items: I,
+    progress: Arc<Mutex<IterProgress>>,
     io: &'io IO,
     r#type: ProgressType,
+    peeked: Option<T>,
 }
 
-impl<'io, T: Debug, IO: Io> CliLogging<'io, T, IO> {
-    fn new<I>(items: I, io: &'io IO, r#type: ProgressType) -> Self
-    where
-        I: IntoIterator<Item = T>,
-    {
-        let items: Vec<_> = items.into_iter().collect();
+impl<'io, T, I, IO> LoggingIterator<'io, T, I, IO>
+where
+    T: Debug,
+    I: Iterator<Item = T>,
+    IO: Io,
+{
+    fn new(
+        items: I,
+        io: &'io IO,
+        r#type: ProgressType,
+        progress: Arc<Mutex<IterProgress>>,
+    ) -> Self {
+        let (size, _) = items.size_hint();
```

Comment:
1. the first thing is a lower bound, does that matter in this context?
2. does it actually work? It looks like the `flume::Receiver` that we wrap in this doesn't override the default impl (which is `(0, None)`)

Reply:
the `size_hint` function implemented on exact sized iterators (like ranges) is always the proper size of the iterator. we wrap two iterators with this `ProgressTracker` abstraction: one that is a [`Range`](https://doc.rust-lang.org/nightly/std/ops/struct.Range.html) and another that is our custom `FetchQueueReceiver` type. the latter implements `size_hint` based on the number of items [that are left to consume on the `flume` channel](https://docs.rs/flume/latest/flume/struct.Receiver.html#method.len). the values returned by this implementation will be a bit racy, but never 0, unless the channel is empty. in the end, we only care about giving some notion of progress to CLI users.